compiler/rustc_codegen_llvm/src/asm.rs

```diff
@@ -874,3 +903,35 @@ fn llvm_fixup_output_type(
         _ => layout.llvm_type(cx),
     }
 }
+
+// Valid llvm symbols: https://llvm.org/docs/AMDGPUOperandSyntax.html#symbols
+// First char is [a-zA-Z_.]
+fn llvm_ident_start(c: char) -> bool {
+    ('a'..='z').contains(&c) || ('A'..='Z').contains(&c) || '_' == c || '.' == c
+}
+
+// All subsequent chars are [a-zA-Z0-9_$.@]
+fn llvm_ident_continuation(c: char) -> bool {
+    ('a'..='z').contains(&c)
+        || ('A'..='Z').contains(&c)
+        || ('0'..='9').contains(&c)
+        || '_' == c
+        || '$' == c
+        || '.' == c
+        || '@' == c
+}
+
+/// Gets a LLVM label from a &str if it exists
+fn get_llvm_label_from_str(s: &str) -> Option<&str> {
+    if let Some(colon_idx) = s.find(':') {
+        let substr = &s[..colon_idx];
+        let mut chars = substr.chars();
+        if let Some(c) = chars.next() {
+            if llvm_ident_start(c) && chars.all(llvm_ident_continuation) {
+                return Some(substr);
+            }
+        }
+    }
```

Comment:
One way to get more arbitrary stuff into a asm label is to quote it as a string, like this:

```c
void test(void) {
    __asm__(".intel_syntax noprefix\n"
    "\"01☺☺@banana.?@!\": jmp \"01☺☺@banana.?@!\"");
}
```

or in Rust

```rust
#![feature(asm)]
pub fn test() {
    unsafe {
    asm!(r#"
        "☺helloworld?@!$%^&": jmp "☺helloworld?@!$%^&"
    "#);
    }
}
```

So that's something we should handle or forbid too.